volatility3/framework/symbols/linux/__init__.py

```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
         kernel = context.modules[kernel_module_name]
 
         return kernel
+
+    @classmethod
+    def choose_kernel_tree(cls, vmlinux: interfaces.context.ModuleInterface) -> "Tree":
+        """Returns the appropriate tree data structure instance for the current kernel implementation.
+        This is used by the IDR and the PageCache to choose between the XArray and RadixTree.
+
+        Args:
+            vmlinux: The kernel module object
+
+        Returns:
+            The appropriate Tree instance for the current kernel
+        """
+        address_space_type = vmlinux.get_type("address_space")
+        address_space_has_i_pages = address_space_type.has_member("i_pages")
+        i_pages_type_name = (
+            address_space_type.child_template("i_pages").vol.type_name
+            if address_space_has_i_pages
+            else ""
+        )
+        i_pages_is_xarray = i_pages_type_name.endswith(constants.BANG + "xarray")
+        i_pages_is_radix_tree_root = i_pages_type_name.endswith(
+            constants.BANG + "radix_tree_root"
+        ) and vmlinux.get_type("radix_tree_root").has_member("xa_head")
+
+        if i_pages_is_xarray or i_pages_is_radix_tree_root:
+            return XArray(vmlinux)
+        else:
+            return RadixTree(vmlinux)
+
+
+class Tree(ABC):
+    """Abstraction to support both XArray and RadixTree"""
+
+    # Dynamic values, these will be initialized later
+    CHUNK_SHIFT = None
+    CHUNK_SIZE = None
+    CHUNK_MASK = None
+
+    def __init__(self, vmlinux: interfaces.context.ModuleInterface):
```

Comment:
For something like this, I'd prefer to pass in the context and a name lookup for `context.modules` please.  It ensures that the state is front and center, and that the rest of the parameters are simple in case they need to be pickled or anything for multithreading down the line.

Reply:
done